File: src-tauri/src/platform/macos.js
// Arguments are JSON data, never interpolated into executable script source.
ObjC.import('ApplicationServices');
function run(argv) {
    var r = JSON.parse(argv[0]);
    var se = Application('System Events');
    if (r.op === 'check') {
        if (!$.AXIsProcessTrusted()) {
            throw new Error('Allow Agent Pulse (and osascript if requested) in Privacy & Security > Accessibility and Automation > System Events, then retry.');
        }
        se.applicationProcesses.length;
        return 'ok';
    }
    if (r.op === 'list') {
        var result = [];
        se.applicationProcesses.whose({backgroundOnly:false})().forEach(function(p) {
            try {
                // One target per window so the Rust resolver reports ambiguity
                // for multiple matching windows, including a persisted PID hint.
                var pid = p.unixId();
                p.windows().forEach(function(w) {
                    result.push({hwnd:pid, process_id:pid, process_name:p.name(),
                        executable_path:null, title:w.name(), is_visible:true});
                });
            } catch (_) { /* inaccessible processes are not targetable */ }
        });
        return JSON.stringify(result);
    }
    var processes = se.applicationProcesses.whose({unixId:r.pid})();
    if (processes.length !== 1) return 'LOST_FOCUS';
    var p = processes[0];
    function guard() {
        if (p.windows().length !== 1) return 'AMBIGUOUS';
        if (!p.frontmost()) return 'LOST_FOCUS';
        // Shift, Control, Option, Command. Leave user-held keys untouched.
        if ((Number($.CGEventSourceFlagsState(1)) & 0x1e0000) !== 0) return 'MODIFIER_HELD';
        return null;
    }
    if (r.op === 'activate') {
        if (p.windows().length !== 1) return 'AMBIGUOUS';
        p.visible = true;
        try { p.windows[0].attributes.byName('AXMinimized').value = false; } catch (_) {}
        p.frontmost = true;
        for (var n = 0; n < 10 && !p.frontmost(); n++) delay(0.05);
        return p.frontmost() ? 'ok' : 'LOST_FOCUS';
    }
    if (r.op === 'foreground') return String(p.frontmost() && p.windows().length === 1);
    var error = guard();
    if (error) return error;
    if (r.op === 'text') {
        // Preserve surrogate pairs; verify focus before every character.
        for (var i = 0; i < r.value.length; i++) {
            error = guard(); if (error) return error;
            var c = r.value.charAt(i);
            var code = r.value.charCodeAt(i);
            if (code >= 0xd800 && code <= 0xdbff && i + 1 < r.value.length) c += r.value.charAt(++i);
            se.keystroke(c);
        }
    } else if (r.op === 'key') {
        if (r.value.code !== undefined) se.keyCode(r.value.code, {using:r.value.modifiers});
        else se.keystroke(r.value.text, {using:r.value.modifiers});
    }
    return 'ok';
}

File: src-tauri/src/platform/macos.rs
//! macOS Accessibility automation through System Events. Each supported target
//! must have exactly one accessible window; multi-window apps fail as ambiguous.
use super::process::run;
use crate::error::{AppError, AppResult};
use crate::executor::PlatformAutomation;
use crate::model::{Key, WindowCandidate};

pub struct MacAutomation;
const SCRIPT: &str = include_str!("macos.js");
fn invoke(op: &str, pid: u64, value: serde_json::Value) -> AppResult<String> {
    let request = serde_json::json!({"op":op,"pid":pid,"value":value}).to_string();
    let result = run(
        "/usr/bin/osascript",
        &["-l", "JavaScript", "-e", SCRIPT, &request],
    )?;
    if result == "LOST_FOCUS" {
        return Err(AppError::TargetLostFocus);
    }
    if result == "AMBIGUOUS" {
        return Err(AppError::TargetAmbiguous);
    }
    if result == "MODIFIER_HELD" {
        return Err(AppError::ModifierKeyHeld);
    }
    Ok(result)
}
fn key_code(key: &Key) -> AppResult<u16> {
    Ok(match key {
        Key::Enter => 36,
        Key::Tab => 48,
        Key::Space => 49,
        Key::Backspace => 51,
        Key::Escape => 53,
        Key::Delete => 117,
        Key::Home => 115,
        Key::End => 119,
        Key::PageUp => 116,
        Key::PageDown => 121,
        Key::Left => 123,
        Key::Right => 124,
        Key::Down => 125,
        Key::Up => 126,
        Key::Function(n) if (1..=12).contains(n) => {
            [122, 120, 99, 118, 96, 97, 98, 100, 101, 109, 103, 111][usize::from(*n - 1)]
        }
        _ => {
            return Err(AppError::InvalidAction(
                "key has no macOS hardware code".into(),
            ));
        }
    })
}
fn key_value(key: &Key, modifiers: &[&str]) -> AppResult<serde_json::Value> {
    match key {
        Key::Letter(c) => Ok(serde_json::json!({"text":c.to_string(),"modifiers":modifiers})),
        Key::Digit(d) => Ok(serde_json::json!({"text":d.to_string(),"modifiers":modifiers})),
        _ => Ok(serde_json::json!({"code":key_code(key)?,"modifiers":modifiers})),
    }
}
impl PlatformAutomation for MacAutomation {
    fn check_available(&self) -> AppResult<()> {
        invoke("check", 0, serde_json::Value::Null)?;
        Ok(())
    }
    fn enumerate(&self) -> Vec<WindowCandidate> {
        invoke("list", 0, serde_json::Value::Null)
            .ok()
            .and_then(|s| serde_json::from_str(&s).ok())
            .unwrap_or_default()
    }
    fn verify_cached(&self, hwnd: u64) -> Option<WindowCandidate> {
        self.enumerate().into_iter().find(|w| w.hwnd == hwnd)
    }
    fn restore(&self, _hwnd: u64) {} // Activation unminimizes the sole window.
    fn activate(&self, hwnd: u64) -> AppResult<()> {
        invoke("activate", hwnd, serde_json::Value::Null)?;
        if self.is_foreground(hwnd) {
            Ok(())
        } else {
            Err(AppError::FailedToActivateTarget)
        }
    }
    fn is_foreground(&self, hwnd: u64) -> bool {
        invoke("foreground", hwnd, serde_json::Value::Null).is_ok_and(|s| s == "true")
    }
    fn type_text(&self, hwnd: u64, text: &str) -> AppResult<()> {
        // Keep helper calls short; the script checks focus before each scalar.
        for chunk in text.chars().collect::<Vec<_>>().chunks(32) {
            invoke(
                "text",
                hwnd,
                serde_json::json!(chunk.iter().collect::<String>()),
            )?;
        }
        Ok(())
    }
    fn press_key(&self, hwnd: u64, key: Key, count: u32, interval_ms: u64) -> AppResult<()> {
        for i in 0..count {
            invoke("key", hwnd, key_value(&key, &[])?)?;
            if i + 1 < count {
                std::thread::sleep(std::time::Duration::from_millis(interval_ms));
            }
        }
        Ok(())
    }
    fn press_combination(&self, hwnd: u64, keys: &[Key]) -> AppResult<()> {
        let modifiers: Vec<&str> = keys
            .iter()
            .filter_map(|k| match k {
                Key::Ctrl => Some("control down"),
                Key::Shift => Some("shift down"),
                Key::Alt => Some("option down"),
                _ => None,
            })
            .collect();
        let others: Vec<&Key> = keys.iter().filter(|k| k.as_modifier().is_none()).collect();
        // System Events supports modifiers + one ordinary key, not held chords.
        if others.len() != 1 {
            return Err(AppError::InvalidAction(
                "macOS combinations require exactly one non-modifier key".into(),
            ));
        }
        invoke("key", hwnd, key_value(others[0], &modifiers)?)?;
        Ok(())
    }
}
